app/controllers/application_controller.rb

```diff
@@ -5,6 +5,143 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "secret"
   end
 
-end
\ No newline at end of file
+  get '/' do
+    erb :index
+  end
+
+  get '/tweets' do
+    # binding.pry
+    if logged_in?
+      @tweets = Tweet.all
+      @user = current_user unless current_user == nil
+      erb :'tweets/tweets'
+    else
+      redirect "/login"
+    end
+  end
+
+  get '/tweets/new' do
+    if logged_in?
+      erb :'tweets/create_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets/new' do
+    if params[:content] != ""
+      @tweet = Tweet.create(:content =>params[:content])
+      @tweet.user = current_user
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    end
+  end
+
+  get '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'tweets/show_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'tweets/edit_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets/:id/edit' do
+    @tweet = Tweet.find_by_id(params[:id])
+    if @tweet.user == current_user && params[:content] != ""
+      @tweet.content = params[:content]
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect "/tweets/#{@tweet.id}/edit"
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+    @tweet = Tweet.find(params[:id])
+    if @tweet.user == current_user
+      @tweet.destroy
+      redirect "/tweets"
+    else
+      redirect "/tweets/#{@tweet.id}"
+    end
+  end
+
+  get '/signup' do
+    # binding.pry
+    if logged_in?
+
+      redirect "/tweets"
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    user = User.new(:username => params[:username], :email => params[:email], :password => params[:password])
+    if user.save && user.username != "" && user.email != "" && user.password != ""
+      session[:id] = user.id
+      redirect "/tweets"
+    else
+      redirect "/signup"
+
+    end
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :'users/login'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(:username => params[:username])
+    if user && user.authenticate(params[:password])
+      session[:id] = user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get "/logout" do
+    if logged_in?
+      session.clear
+      redirect "/login"
+    else
+      redirect "/login"
+    end
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/user_tweets'
+    # binding.pry
+  end
+
+  helpers do
+    def logged_in?
+      !!session[:id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```